### ISSUE TITLE
Mean Field Dev - dynamics.py

### ISSUE THREAD

**Issue by gefux** [tempoCollaboration/OQuPy#55]
Hi @piperfw !
As we've discussed the `dynamics.py` module should probably get a new structure. Here my suggestion ...

```python
class BaseDynamics(BaseAPIClass):
    def __init__(
            self,
            name: Optional[Text] = None,
            description: Optional[Text] = None) -> None:
        # This class doesn't have a method to populate times and states
        # but it has the method to return what's there and to
        # compute expectations
        super().__init__(name, description)
        self._times = []
        self._states = []

    def __str__(self) -> Text:
        pass # ToDo

    def __len__(self) -> int:
        pass # ToDo

    @property
    def times(self) -> ndarray:
        pass # ToDo

    @property
    def states(self) -> ndarray:
        pass # ToDo

    @property
    def shape(self) -> ndarray:
        pass # ToDo

    def expectations(
            self,
            operator: Optional[ndarray] = None,
            real: Optional[bool] = False) -> Tuple[ndarray, ndarray]:
        pass # ToDo


class Dynamics(BaseDynamics):
    def __init__(
            self,
            times: Optional[List[float]] = None,
            states: Optional[List[ndarray]] = None,
            name: Optional[Text] = None,
            description: Optional[Text] = None) -> None:
        super().__init__(name, description)
        for time, state in zip(times, states):
            self.add(time, state)

    def add(
            self,
            time: float,
            state: ndarray) -> None:
        tmp_time = _parse_time(time)
        tmp_state = _parse_state(state)
        index = _find_list_index(self._times, tmp_time)
        self._times.insert(index, tmp_time)
        self._states.insert(index, tmp_state)

class DynamicsWithField(BaseDynamics):
    def __init__(
            self,
            times: Optional[List[float]] = None,
            states: Optional[List[ndarray]] = None,
            fields: Optional[List[complex]] = None,
            name: Optional[Text] = None,
            description: Optional[Text] = None) -> None:
        super().__init__(name, description)
        self._fields = []
        for time, state, field in zip(times, states, fields):
            self.add(time, state, field)

    @property
    def fields(self) -> ndarray:
        pass # ToDo

    def field_expectations(self) -> Tuple(ndarray, ndarray):
        pass # ToDo

    def add(
            self,
            time: float,
            state: ndarray,
            field: complex) -> None:
        tmp_time = _parse_time(time)
        tmp_state = _parse_state(state)
        tmp_field = _parse_field(field)
        index = _find_list_index(self._times, tmp_time)
        self._times.insert(index, tmp_time)
        self._states.insert(index, tmp_state)
        self._fields.insert(index, tmp_field)

def _parse_time(time) -> float:
    pass # ToDo

def _parse_state(state) -> ndarray:
    pass # ToDo

def _parse_field(field) -> complex:
    pass # ToDo

def _find_list_index(sorted_list, entry_value) -> int:
    """
    Return `index` such that a `sorted_list` stays sorted when extended with
    `entry_value` like so:
       `sorted_list.insert(index, entry_value)`
    """
    pass # ToDo

```

**Comment by piperfw** [comment]
Hi @gefux 

That's great, thanks for the detailed outline. I've made the changes on [piperfw/TimeEvolvingMPO](https://github.com/piperfw/TimeEvolvingMPO/tree/mean-field-dev). Comments:

- I tried to stick closely to the current format of the Dynamics class on oqupy-dev. In particular BaseDynamics also has `export()` and `self._shape` in addition to the above
- `bisect` from the standard library is used to get the right index to insert a time/state/field without sorting `self._times` list
- The possible computational saving from the use of `self._expectation_operators` and `self._expectation_lists` in `add()` hardly seems worth the code complexity (or memory). Left these in for now but would consider removing

**Comment by piperfw** [comment]
Great, thanks for the comments @gefux 
- `export()` and the expectation-saving lists have been removed, along with `_sort()`
- the unnecessary `copy` has been removed and `__operator` renamed to `tmp_operator`
- tests have been updated for `BaseDynamics`, `DynamicsWithField`, and pylint is happy

**Comment by gefux** [comment]
Great. @piperfw: Would you be happy to create a pull request (targeting the `mean-field-dev` branch)?

**Comment by gefux** [comment]
changes are merged into the mean-field-dev branch #62
